test/jdk/java/lang/ProcessBuilder/JspawnhelperProtocol.java

```diff
@@ -0,0 +1,183 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @bug 8307990
+ * @requires (os.family == "linux" & !vm.musl)
+ * @requires vm.debug
+ * @library /test/lib
+ * @run main/othervm/timeout=300 JspawnhelperProtocol
+ */
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import jdk.test.lib.process.ProcessTools;
+
+public class JspawnhelperProtocol {
+    // Timout in seconds
+    private static final int TIMEOUT = 60;
+    // Base error code to communicate various error states from grandchild to child process
+    private static final int ERROR = 10;
+    private static final String[] CMD = { "pwd" };
+    private static final String ENV_KEY = "JTREG_JSPAWNHELPER_PROTOCOL_TEST";
+
+    private static void childCode(String arg) throws IOException, InterruptedException {
+        System.out.println("Recursively executing 'JspawnhelperProtocol " + arg + "'");
+        Process p = null;
+        try {
+            p = Runtime.getRuntime().exec(CMD);
+        } catch (Exception e) {
+            System.out.println(e.toString());
+            System.exit(ERROR);
+        }
+        if (!p.waitFor(TIMEOUT, TimeUnit.SECONDS)) {
+            System.out.println("Grandchild process timed out");
+            System.exit(ERROR + 1);
+        }
+        if (p.exitValue() == 0) {
+            String pwd = p.inputReader().readLine();
+            String realPwd = Path.of("").toAbsolutePath().toString();
+            if (!realPwd.equals(pwd)) {
+                System.out.println("Grandchild process returned '" + pwd + "' (expected '" + realPwd + "')");
+                System.exit(ERROR + 2);
+            }
+            System.out.println("  Successfully executed '" + CMD[0] + "'");
+            System.exit(0);
+        } else {
+            System.out.println("  Failed to executed '" + CMD[0] + "' (exitValue=" + p.exitValue() + ")");
+            System.exit(ERROR + 3);
+        }
+    }
+
+    private static void normalExec() throws Exception {
+        ProcessBuilder pb;
+        pb = ProcessTools.createJavaProcessBuilder("-Djdk.lang.Process.launchMechanism=posix_spawn",
+                                                   "JspawnhelperProtocol",
+                                                   "normalExec");
+        pb.inheritIO();
+        Process p = pb.start();
+        if (!p.waitFor(TIMEOUT, TimeUnit.SECONDS)) {
+            throw new Exception("Child process timed out");
+        }
+        if (p.exitValue() != 0) {
+            throw new Exception("Child process exited with " + p.exitValue());
+        }
+    }
+
```

Comment:
This sounds reasonable. I've renamed `childCode()` to `parentCode()`, streamlined the wording for "parent" and "child" in the various log messages and exceptions and added a comment to the `main()` method which explains the working of the test and the meaning of "parent" and "child" in the output.

I've not merged `simulateCrashInChild()` and `simulateCrashInParent()` because there are some subtle differences between the two (and now with the improved "parent"/"child" naming even more :) and I don't think that a merged version will improve the readability.